core/src/main/java/org/infinispan/transaction/TransactionTable.java

```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
-               long ageNanos = System.nanoTime() - e.getValue();
-               if (TimeUnit.NANOSECONDS.toMillis(ageNanos) >= timeout) {
-                  iterator.remove();
-                  removedEntries++;
-               }
-            }
+        if (completedTransactionsInfo != null && !completedTransactionsInfo.isEmpty()) {
+            try {
+                log.tracef("About to cleanup completed transaction. Initial size is %d", completedTransactionsInfo.size());
+                long beginning = System.nanoTime();
+                long minCompleteTimestamp = System.nanoTime()
+                        - TimeUnit.MILLISECONDS.toNanos(configuration.transaction().completedTxTimeout());
+                int removedEntries = 0;
+
+                // this iterator is weekly consistent and will never throw ConcurrentModificationException
+                Iterator<Map.Entry<Address, CompletedTransactionsPerNode>> nodeIterator = completedTransactionsInfo.entrySet()
+                        .iterator();
+
+                while (nodeIterator.hasNext()) {
+                    CompletedTransactionsPerNode nodeInfo = nodeIterator.next().getValue();
+                    // Prune nodes that are no longer members first
+                    if (nodeInfo.completedTransactions.isEmpty() && !rpcManager.getMembers().contains(nodeInfo.address)) {
+                        nodeIterator.remove();
+                    }
+
+                    // Now prune the individual transactions
+                    Iterator<Map.Entry<GlobalTransaction, Long>> txIterator = nodeInfo.completedTransactions.entrySet()
+                            .iterator();
+                    while (txIterator.hasNext()) {
+                        Map.Entry<GlobalTransaction, Long> e = txIterator.next();
+                        long completedTime = e.getValue();
+                        if (completedTime < minCompleteTimestamp) {
+                            // Need to update lastPrunedTxId *before* removing the tx from the map
+                            // Don't need atomic operations, there can't be more than one thread updating lastPrunedTxId.
+                            long txId = e.getKey().getId();
+                            if (txId > nodeInfo.lastPrunedTxId) {
+                                nodeInfo.lastPrunedTxId = txId;
+                            }
+
+                            txIterator.remove();
+                            removedEntries++;
+                        }
+                    }
+                }
             long duration = System.nanoTime() - beginning;
 
             log.tracef("Finished cleaning up completed transactions. %d transactions were removed, total duration was %d millis, " +
                   "current number of completed transactions is %d", removedEntries, TimeUnit.NANOSECONDS.toMillis(duration),
-                  completedTransactions.size());
+                  completedTransactionsInfo.size());
```

Comment:
I think this is wrong, `completedTransactionsInfo.size()` is just number of nodes, not transactions. The same with the trace log on the beginning.